src/rabbit_connection_tracking.erl

```diff
@@ -0,0 +1,340 @@
+
+register_connection(#tracked_connection{vhost = VHost, id = ConnId, node = Node} = Conn) when Node =:= node() ->
+    TableName = tracked_connection_table_name_for(Node),
+    PerVhostTableName = tracked_connection_per_vhost_table_name_for(Node),
+    rabbit_misc:execute_mnesia_transaction(
+      fun() ->
+        %% upsert
+              case mnesia:dirty_read(TableName, ConnId) of
+                  []    ->
+                    %% TODO: counter table
+                    mnesia:write(TableName, Conn, write),
+                    mnesia:dirty_update_counter(
+                        PerVhostTableName, VHost, 1);
+                  [_Row] ->
+                      ok
+              end,
+              ok
+      end).
+
+-spec unregister_connection(rabbit_types:connection_name()) -> ok.
+
+unregister_connection(ConnId = {Node, _Name}) when Node =:= node() ->
+    TableName = tracked_connection_table_name_for(Node),
+    PerVhostTableName = tracked_connection_per_vhost_table_name_for(Node),
+    rabbit_misc:execute_mnesia_transaction(
+      fun() ->
+              case mnesia:dirty_read(TableName, ConnId) of
+                  []     -> ok;
+                  [Row] ->
+                      mnesia:dirty_update_counter(
+                        PerVhostTableName, Row#tracked_connection.vhost, -1),
+                      mnesia:delete({TableName, ConnId})
+              end
+      end).
+
+
+-spec list() -> [rabbit_types:tracked_connection()].
+
+list() ->
+  Chunks = lists:map(
+    fun (Node) ->
+      Tab = tracked_connection_table_name_for(Node),
+      mnesia:dirty_match_object(Tab, #tracked_connection{_ = '_'})
+    end, rabbit_mnesia:cluster_nodes(running)),
+  lists:foldl(fun(Chunk, Acc) -> Acc ++ Chunk end, [], Chunks).
+
+
+-spec list(rabbit_types:vhost()) -> [rabbit_types:tracked_connection()].
+
+list(VHost) ->
+  Chunks = lists:map(
+    fun (Node) ->
+      Tab = tracked_connection_table_name_for(Node),
+      mnesia:dirty_match_object(Tab, #tracked_connection{vhost = VHost, _ = '_'})
+    end, rabbit_mnesia:cluster_nodes(running)),
+  lists:foldl(fun(Chunk, Acc) -> Acc ++ Chunk end, [], Chunks).
+
+
+-spec list_on_node(node()) -> [rabbit_types:tracked_connection()].
+
+list_on_node(Node) ->
+  try mnesia:dirty_match_object(
+    tracked_connection_table_name_for(Node),
+    #tracked_connection{_ = '_'})
+  catch exit:{aborted, {no_exists, _}} -> []
+  end.
+
+-spec is_over_connection_limit(rabbit_types:vhost()) -> {true, non_neg_integer()} | false.
+
+is_over_connection_limit(VirtualHost) ->
+    case rabbit_vhost_limit:connection_limit(VirtualHost) of
+        %% no limit configured
+        undefined                                            -> false;
+        %% with limit = 0, no connections are allowed
+        {ok, 0}                                              -> {true, 0};
+        {ok, Limit} when is_integer(Limit) andalso Limit > 0 ->
+          ConnectionCount = count_connections_in(VirtualHost),
```

Comment:
on coding style, can we stick to our general indentation rule, of the default Erlang mode for Emacs? your indentation also seems to differ from function to function. for example `boot/0` and `ensure_tracked_connections_table_for_node/1` appear to use different indentation.